### PR TITLE
Pure intersection types support, `never` type support

### LINKED ISSUE #99: PHP 8.1 pure intersection type support
https://wiki.php.net/rfc/pure-intersection-types

### LINKED ISSUE #102: PHP 8.1 noreturn type support
https://wiki.php.net/rfc/noreturn_type

### PR DESCRIPTION
**Pull request by Ocramius**
|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | no
| BC Break      | no
| New Feature   | yes
| RFC           | no
| QA            | no

### Description

Implements pure intersection types (`Foo&Bar`) and the `never` built-in type

Fixes #99
Fixes #102

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -18,18 +18,21 @@
...
     "suggest": {
         "doctrine/annotations": "Doctrine\\Common\\Annotations >=1.0 for annotation features",
         "laminas/laminas-stdlib": "Laminas\\Stdlib component",
         "laminas/laminas-zendframework-bridge": "A bridge with Zend Framework"
     },
     "config": {
-        "sort-packages": true
+        "sort-packages": true,
+        "platform": {
+            "php": "7.4.99"
```
Comment (boesing) [inline composer.json:34]: Just curious on why do we need this?

File: src/Generator/TypeGenerator.php [hunk 2]
```diff
@@ -149,18 +173,18 @@ private function __construct(array $types, bool $nullable)
...
     {
         $typesAsStrings = array_map(
-            fn(AtomicType $type): string => $type->fullyQualifiedName(),
+            static fn (AtomicType $type): string => $type->fullyQualifiedName(),
             $this->types
         );
 
         if ($this->nullable) {
-            return '?' . implode('|', $typesAsStrings);
+            return '?' . implode($this->atomicTypesSeparator(), $typesAsStrings);
```
Comment (boesing) [inline src/Generator/TypeGenerator.php:184]: Intersection types are not nullable as of PHP 8.1, right? Wonder if that should be validated somewhere? Or did I miss something somewhere?

File: src/Generator/TypeGenerator.php [hunk 3]
```diff
@@ -188,12 +212,20 @@ public function __toString()
      * @psalm-return array{bool, string}
```
Comment (boesing) [inline src/Generator/TypeGenerator.php:212]: Uh, wasnt aware that one can write it that way (without the numeric keys). TIL

File: test/Generator/TypeGeneratorTest.php [hunk 4]
```diff
@@ -239,88 +258,149 @@ public function invalidType()
             ['1foo'],
             ['foo\\1foo'],
             ['?foo\\bar|null'],
+            ['?foo\\bar&null'],
```
Comment (boesing) [inline test/Generator/TypeGeneratorTest.php:261]: intersection types are not nullable, arent they?

File: test/Generator/TypeGeneratorTest.php [hunk 5]
```diff
@@ -239,88 +258,149 @@ public function invalidType()
...
             ['*'],
             ["\0"],
             ['\\array'],
             ['\\Array'],
             ['\\ARRAY'],
             ['\\array|null'],
+            ['\\array&null'],
             ['null|\\array'],
+            ['null&\\array'],
```
Comment (boesing) [inline test/Generator/TypeGeneratorTest.php:270]: intersection types are not nullable, arent they?

File: test/Generator/TypeGeneratorTest.php [hunk 6]
```diff
@@ -239,88 +258,149 @@ public function invalidType()
...
             ['?array|null'],
+            ['?array&null'],
```
Comment (boesing) [inline test/Generator/TypeGeneratorTest.php:272]: intersection types are not nullable, arent they?

File: test/Generator/TypeGeneratorTest.php [hunk 7]
```diff
@@ -239,88 +258,149 @@ public function invalidType()
...
             ['\\callable'],
             ['\\Callable'],
             ['\\CALLABLE'],
             ['\\callable|null'],
+            ['\\callable&null'],
```
Comment (boesing) [inline test/Generator/TypeGeneratorTest.php:277]: intersection types are not nullable, arent they?

File: test/Generator/TypeGeneratorTest.php [hunk 8]
```diff
@@ -239,88 +258,149 @@ public function invalidType()
...
             ['null|\\callable'],
+            ['null&\\callable'],
```
Comment (boesing) [inline test/Generator/TypeGeneratorTest.php:279]: intersection types are not nullable, arent they?

File: test/Generator/TypeGeneratorTest.php [hunk 9]
```diff
@@ -239,88 +258,149 @@ public function invalidType()
...
             ['?callable|null'],
+            ['?callable&null'],
```
Comment (boesing) [inline test/Generator/TypeGeneratorTest.php:281]: intersection types are not nullable, arent they?

File: test/Generator/TypeGeneratorTest.php [hunk 10]
```diff
@@ -239,88 +258,149 @@ public function invalidType()
...
             ['\\string'],
             ['\\String'],
             ['\\STRING'],
             ['\\string|null'],
+            ['\\string&null'],
```
Comment (boesing) [inline test/Generator/TypeGeneratorTest.php:286]: intersection types are not nullable, arent they?

File: test/Generator/TypeGeneratorTest.php [hunk 11]
```diff
@@ -239,88 +258,149 @@ public function invalidType()
...
             ['null|\\string'],
+            ['null&\\string'],
```
Comment (boesing) [inline test/Generator/TypeGeneratorTest.php:288]: intersection types are not nullable, arent they?

File: test/Generator/TypeGeneratorTest.php [hunk 12]
```diff
@@ -239,88 +258,149 @@ public function invalidType()
...
             ['?string|null'],
+            ['?string&null'],
```
Comment (boesing) [inline test/Generator/TypeGeneratorTest.php:290]: intersection types are not nullable, arent they?

File: test/Generator/TypeGeneratorTest.php [hunk 13]
```diff
@@ -239,88 +258,149 @@ public function invalidType()
...
             ['\\int'],
             ['\\Int'],
             ['\\INT'],
             ['\\int|null'],
+            ['\\int&null'],
```
Comment (boesing) [inline test/Generator/TypeGeneratorTest.php:295]: intersection types are not nullable, arent they?

File: test/Generator/TypeGeneratorTest.php [hunk 14]
```diff
@@ -239,88 +258,149 @@ public function invalidType()
...
             ['null|\\int'],
+            ['null&\\int'],
```
Comment (boesing) [inline test/Generator/TypeGeneratorTest.php:297]: intersection types are not nullable, arent they?

File: test/Generator/TypeGeneratorTest.php [hunk 15]
```diff
@@ -239,88 +258,149 @@ public function invalidType()
...
             ['?int|null'],
+            ['?int&null'],
```
Comment (boesing) [inline test/Generator/TypeGeneratorTest.php:299]: intersection types are not nullable, arent they?

File: test/Generator/TypeGeneratorTest.php [hunk 16]
```diff
@@ -239,88 +258,149 @@ public function invalidType()
...
             ['\\float'],
             ['\\Float'],
             ['\\FLOAT'],
             ['\\float|null'],
+            ['\\float&null'],
```
Comment (boesing) [inline test/Generator/TypeGeneratorTest.php:304]: intersection types are not nullable, arent they?

File: test/Generator/TypeGeneratorTest.php [hunk 17]
```diff
@@ -239,88 +258,149 @@ public function invalidType()
...
             ['null|\\float'],
+            ['null&\\float'],
```
Comment (boesing) [inline test/Generator/TypeGeneratorTest.php:306]: intersection types are not nullable, arent they?

File: test/Generator/TypeGeneratorTest.php [hunk 18]
```diff
@@ -239,88 +258,149 @@ public function invalidType()
...
             ['?float|null'],
+            ['?float&null'],
```
Comment (boesing) [inline test/Generator/TypeGeneratorTest.php:308]: intersection types are not nullable, arent they?

File: test/Generator/TypeGeneratorTest.php [hunk 19]
```diff
@@ -239,88 +258,149 @@ public function invalidType()
...
             ['\\false'],
             ['\\FALSE'],
             ['\\False'],
             ['?false|null'],
+            ['?false&null'],
```
Comment (boesing) [inline test/Generator/TypeGeneratorTest.php:313]: intersection types are not nullable, arent they?

File: src/Generator/TypeGenerator.php (hunk 2 above)
Comment (boesing) [reply src/Generator/TypeGenerator.php:184]: https://3v4l.org/pO7JhB#v8.1rc2

File: test/Generator/TypeGeneratorTest.php [hunk 20]
```diff
@@ -239,88 +258,149 @@ public function invalidType()
...
             ['\\bool'],
             ['\\Bool'],
             ['\\BOOL'],
             ['\\bool|null'],
+            ['\\bool&null'],
```
Comment (boesing) [inline test/Generator/TypeGeneratorTest.php:318]: intersection types are not nullable, arent they?

File: test/Generator/TypeGeneratorTest.php [hunk 21]
```diff
@@ -239,88 +258,149 @@ public function invalidType()
...
             ['null|\\bool'],
+            ['null&\\bool'],
```
Comment (boesing) [inline test/Generator/TypeGeneratorTest.php:320]: intersection types are not nullable, arent they?

File: test/Generator/TypeGeneratorTest.php [hunk 22]
```diff
@@ -239,88 +258,149 @@ public function invalidType()
...
             ['?bool|null'],
+            ['?bool&null'],
```
Comment (boesing) [inline test/Generator/TypeGeneratorTest.php:322]: intersection types are not nullable, arent they?

File: test/Generator/TypeGeneratorTest.php [hunk 23]
```diff
@@ -239,88 +258,149 @@ public function invalidType()
...
             ['\\void'],
             ['\\Void'],
             ['\\VOID'],
             ['\\void|null'],
+            ['\\void&null'],
```
Comment (boesing) [inline test/Generator/TypeGeneratorTest.php:327]: intersection types are not nullable, arent they?

File: test/Generator/TypeGeneratorTest.php [hunk 24]
```diff
@@ -239,88 +258,149 @@ public function invalidType()
...
             ['null|\\void'],
+            ['null&\\void'],
```
Comment (boesing) [inline test/Generator/TypeGeneratorTest.php:329]: intersection types are not nullable, arent they?

File: test/Generator/TypeGeneratorTest.php [hunk 25]
```diff
@@ -239,88 +258,149 @@ public function invalidType()
...
             ['?void'],
             ['?Void'],
             ['?VOID'],
             ['void|null'],
+            ['void&null'],
```
Comment (boesing) [inline test/Generator/TypeGeneratorTest.php:334]: intersection types are not nullable, arent they?

File: test/Generator/TypeGeneratorTest.php [hunk 26]
```diff
@@ -239,88 +258,149 @@ public function invalidType()
...
             ['?void|null'],
+            ['?void&null'],
```
Comment (boesing) [inline test/Generator/TypeGeneratorTest.php:336]: intersection types are not nullable, arent they?

File: test/Generator/TypeGeneratorTest.php [hunk 27]
```diff
@@ -239,88 +258,149 @@ public function invalidType()
...
             ['\\iterable'],
             ['\\Iterable'],
             ['\\ITERABLE'],
             ['\\iterable|null'],
+            ['\\iterable&null'],
```
Comment (boesing) [inline test/Generator/TypeGeneratorTest.php:341]: intersection types are not nullable, arent they?

File: test/Generator/TypeGeneratorTest.php [hunk 28]
```diff
@@ -239,88 +258,149 @@ public function invalidType()
...
             ['null|\\iterable'],
+            ['null&\\iterable'],
```
Comment (boesing) [inline test/Generator/TypeGeneratorTest.php:343]: intersection types are not nullable, arent they?

File: test/Generator/TypeGeneratorTest.php [hunk 29]
```diff
@@ -239,88 +258,149 @@ public function invalidType()
...
             ['?iterable|null'],
+            ['?iterable&null'],
```
Comment (boesing) [inline test/Generator/TypeGeneratorTest.php:345]: intersection types are not nullable, arent they?

File: test/Generator/TypeGeneratorTest.php [hunk 30]
```diff
@@ -239,88 +258,149 @@ public function invalidType()
...
             ['\\object'],
             ['\\Object'],
             ['\\OBJECT'],
             ['\\object|null'],
+            ['\\object&null'],
```
Comment (boesing) [inline test/Generator/TypeGeneratorTest.php:350]: intersection types are not nullable, arent they?

File: test/Generator/TypeGeneratorTest.php [hunk 31]
```diff
@@ -239,88 +258,149 @@ public function invalidType()
...
             ['null|\\object'],
+            ['null&\\object'],
```
Comment (boesing) [inline test/Generator/TypeGeneratorTest.php:352]: intersection types are not nullable, arent they?

File: test/Generator/TypeGeneratorTest.php [hunk 32]
```diff
@@ -239,88 +258,149 @@ public function invalidType()
...
             ['?object|null'],
+            ['?object&null'],
```
Comment (boesing) [inline test/Generator/TypeGeneratorTest.php:354]: intersection types are not nullable, arent they?

File: test/Generator/TypeGeneratorTest.php [hunk 33]
```diff
@@ -239,88 +258,149 @@ public function invalidType()
...
+
+            // `mixed` can not be intersect-ed with anything
+            ['?mixed'],
+            ['mixed&null'],
+            ['mixed&Foo'],
+            ['mixed&\\foo'],
+
+            // `never` can not be intersect-ed with anything
+            ['?never'],
+            ['never&null'],
```
Comment (boesing) [inline test/Generator/TypeGeneratorTest.php:400]: intersection types are not nullable, arent they?

File: test/Generator/TypeGeneratorTest.php [hunk 34]
```diff
@@ -239,88 +258,149 @@ public function invalidType()
...
 
+            // `never` can not be union-ed with anything
+            ['?never'],
+            ['never|null'],
+            ['never|Foo'],
+            ['never|\\foo'],
+
+            // `mixed` can not be intersect-ed with anything
+            ['?mixed'],
+            ['mixed&null'],
```
Comment (boesing) [inline test/Generator/TypeGeneratorTest.php:394]: intersection types are not nullable, arent they?

File: test/Generator/TypeGeneratorTest.php [hunk 35]
```diff
@@ -239,88 +258,149 @@ public function invalidType()
...
+            ['\\mixed&null'],
             ['null|\\mixed'],
+            ['null&\\mixed'],
+            ['\\never'],
+            ['\\NEVER'],
+            ['\\Never'],
+            ['\\never|null'],
+            ['\\never&null'],
+            ['null|\\never'],
+            ['null&\\never'],
```
Comment (boesing) [inline test/Generator/TypeGeneratorTest.php:378]: intersection types are not nullable, arent they?

File: test/Generator/TypeGeneratorTest.php [hunk 36]
```diff
@@ -239,88 +258,149 @@ public function invalidType()
...
             ['\\Mixed'],
             ['\\mixed|null'],
+            ['\\mixed&null'],
             ['null|\\mixed'],
+            ['null&\\mixed'],
+            ['\\never'],
+            ['\\NEVER'],
+            ['\\Never'],
+            ['\\never|null'],
+            ['\\never&null'],
```
Comment (boesing) [inline test/Generator/TypeGeneratorTest.php:376]: intersection types are not nullable, arent they?

File: test/Generator/TypeGeneratorTest.php [hunk 37]
```diff
@@ -239,88 +258,149 @@ public function invalidType()
...
+            ['null&\\static'],
             ['?static|null'],
+            ['?static&null'],
             ['\\mixed'],
             ['\\MIXED'],
             ['\\Mixed'],
             ['\\mixed|null'],
+            ['\\mixed&null'],
             ['null|\\mixed'],
+            ['null&\\mixed'],
```
Comment (boesing) [inline test/Generator/TypeGeneratorTest.php:371]: intersection types are not nullable, arent they?

File: test/Generator/TypeGeneratorTest.php [hunk 38]
```diff
@@ -239,88 +258,149 @@ public function invalidType()
...
+            ['static&\\foo'],
             ['null|\\static'],
+            ['null&\\static'],
             ['?static|null'],
+            ['?static&null'],
             ['\\mixed'],
             ['\\MIXED'],
             ['\\Mixed'],
             ['\\mixed|null'],
+            ['\\mixed&null'],
```
Comment (boesing) [inline test/Generator/TypeGeneratorTest.php:369]: intersection types are not nullable, arent they?

File: test/Generator/TypeGeneratorTest.php [hunk 39]
```diff
@@ -239,88 +258,149 @@ public function invalidType()
...
             ['\\static'],
             ['\\STATIC'],
             ['\\Static'],
             ['\\static|null'],
+            ['\\static&null'],
+            ['static&\\foo'],
             ['null|\\static'],
+            ['null&\\static'],
             ['?static|null'],
+            ['?static&null'],
```
Comment (boesing) [inline test/Generator/TypeGeneratorTest.php:364]: intersection types are not nullable, arent they?

File: test/Generator/TypeGeneratorTest.php [hunk 40]
```diff
@@ -239,88 +258,149 @@ public function invalidType()
...
             ['?object|null'],
+            ['?object&null'],
             ['\\static'],
             ['\\STATIC'],
             ['\\Static'],
             ['\\static|null'],
+            ['\\static&null'],
+            ['static&\\foo'],
             ['null|\\static'],
+            ['null&\\static'],
```
Comment (boesing) [inline test/Generator/TypeGeneratorTest.php:362]: intersection types are not nullable, arent they?

File: test/Generator/TypeGeneratorTest.php [hunk 41]
```diff
@@ -239,88 +258,149 @@ public function invalidType()
...
+            ['\\object&null'],
             ['null|\\object'],
+            ['null&\\object'],
             ['?object|null'],
+            ['?object&null'],
             ['\\static'],
             ['\\STATIC'],
             ['\\Static'],
             ['\\static|null'],
+            ['\\static&null'],
```
Comment (boesing) [inline test/Generator/TypeGeneratorTest.php:359]: intersection types are not nullable, arent they?

**Review by boesing — CHANGES_REQUESTED** [review]

File: composer.json (hunk 1 above)
Comment (Ocramius) [reply composer.json:34]: If you don't do this, a `composer update` from any `8.0` or `8.1` container (I use docker containers for developing) will cause a dependency bump to something incompatible with PHP 7.4.

That's visible in `git blame`

File: src/Generator/TypeGenerator.php (hunk 2 above)
Comment (Ocramius) [reply src/Generator/TypeGenerator.php:184]: Yeah, they are not nullable, but this entire `TypeGenerator` is quite lax about inputs, as its only job is really to just produce a new type.

In fact, union types are also quite wonky: https://3v4l.org/tkNtB#v8.0.11

The `TypeGenerator` still works with `?int|float` though, so the flaw is pre-existing

File: test/Generator/TypeGeneratorTest.php (hunk 4 above)
Comment (Ocramius) [reply test/Generator/TypeGeneratorTest.php:261]: This is in `invalidType()` :+1:

File: test/Generator/TypeGeneratorTest.php (hunk 6 above)
Comment (Ocramius) [reply test/Generator/TypeGeneratorTest.php:272]: This is in `invalidType()` :+1:

File: test/Generator/TypeGeneratorTest.php (hunk 5 above)
Comment (Ocramius) [reply test/Generator/TypeGeneratorTest.php:270]: This is in `invalidType()` :+1:

File: test/Generator/TypeGeneratorTest.php (hunk 7 above)
Comment (Ocramius) [reply test/Generator/TypeGeneratorTest.php:277]: This is in `invalidType()` :+1:

File: test/Generator/TypeGeneratorTest.php (hunk 9 above)
Comment (Ocramius) [reply test/Generator/TypeGeneratorTest.php:281]: This is in `invalidType()` :+1:

File: test/Generator/TypeGeneratorTest.php (hunk 10 above)
Comment (Ocramius) [reply test/Generator/TypeGeneratorTest.php:286]: This is in `invalidType()` :+1:

File: test/Generator/TypeGeneratorTest.php (hunk 12 above)
Comment (Ocramius) [reply test/Generator/TypeGeneratorTest.php:290]: This is in `invalidType()` :+1:

File: test/Generator/TypeGeneratorTest.php (hunk 14 above)
Comment (Ocramius) [reply test/Generator/TypeGeneratorTest.php:297]: This is in `invalidType()` :+1:

File: test/Generator/TypeGeneratorTest.php (hunk 16 above)
Comment (Ocramius) [reply test/Generator/TypeGeneratorTest.php:304]: This is in `invalidType()` :+1:

File: test/Generator/TypeGeneratorTest.php (hunk 18 above)
Comment (Ocramius) [reply test/Generator/TypeGeneratorTest.php:308]: This is in `invalidType()` :+1:

Comment (Ocramius) [reply test/Generator/TypeGeneratorTest.php:308]: I dismissed further similar feedback on the `invalidType()` data provider

**Review by boesing — APPROVED** [review]
Looks reasonable.

**Comment by Ocramius** [comment]
:ship: then :D

**Pull request merged**